setup.py

```diff
@@ -13,12 +14,173 @@
+            "-DCMAKE_BUILD_TYPE={}".format(build_type),
+            "-DPYTHON_EXECUTABLE={}".format(sys.executable),
+            "-DUSE_PYTHON_CASADI={}".format("TRUE" if use_python_casadi else "FALSE"),
+        ]
+        if self.suitesparse_root:
+            cmake_args.append(
+                "-DSuiteSparse_ROOT={}".format(os.path.abspath(self.suitesparse_root))
+            )
+        if self.sundials_root:
+            cmake_args.append(
+                "-DSUNDIALS_ROOT={}".format(os.path.abspath(self.sundials_root))
+            )
+
+        build_dir = self.get_build_directory()
+        if not os.path.exists(build_dir):
+            os.makedirs(build_dir)
+
+        # The CMakeError.log file is generated by cmake is the configure step
+        # encounters error. In the following the existence of this file is used
+        # to determine whether or not the cmake configure step went smoothly.
+        # So must make sure this file does not remain from a previous failed build.
+        if os.path.isfile(os.path.join(build_dir, "CMakeError.log")):
+            os.remove(os.path.join(build_dir, "CMakeError.log"))
+
+        build_env = os.environ
+        if os.getenv("PYBAMM_USE_VCPKG"):
+            (
+                vcpkg_root_dir,
+                vcpkg_default_triplet,
+                vcpkg_feature_flags,
+            ) = set_vcpkg_environment_variables()
+            build_env["vcpkg_root_dir"] = vcpkg_root_dir
+            build_env["vcpkg_default_triplet"] = vcpkg_default_triplet
+            build_env["vcpkg_feature_flags"] = vcpkg_feature_flags
+
+        cmake_list_dir = os.path.abspath(os.path.dirname(__file__))
+        print("-" * 10, "Running CMake for idaklu solver", "-" * 40)
+        subprocess.run(
+            ["cmake", cmake_list_dir] + cmake_args, cwd=build_dir, env=build_env
+        )
+
+        if os.path.isfile(os.path.join(build_dir, "CMakeError.log")):
+            msg = (
+                "cmake configuration steps encountered errors, and the idaklu module"
+                " could not be built. Make sure dependencies are correctly "
+                "installed. See "
+                "https://github.com/pybamm-team/PyBaMM/tree/develop"
+                "INSTALL-LINUX-MAC.md"
+            )
+            raise RuntimeError(msg)
+        else:
+            print("-" * 10, "Building idaklu module", "-" * 40)
+            subprocess.run(
+                ["cmake", "--build", ".", "--config", "Release"],
+                cwd=build_dir,
+                env=build_env,
+            )
+
+            # Move from build temp to final position
+            for ext in self.extensions:
+                self.move_output(ext)
+
+    def move_output(self, ext):
+        # Copy built module to dist/ directory
+        build_temp = Path(self.build_temp).resolve()
+        # Get destination location
+        # self.get_ext_fullpath(ext.name) -->
+        # build/lib.linux-x86_64-3.5/idaklu.cpython-37m-x86_64-linux-gnu.so
```

Comment:
It is possible that the error with the build isolation came from here, and we can remove `casadi` from the build-time dependencies if this functionality is modified.

Reply:
So from what I can see when the wheels are repaired on macos and linux it can't find the casadi lib. In the CMake we execute a bit of python to find the exact directory (in site-packages) that casadi is in ${CASADI_DIR}, then link against this. 

When the wheel is repaired, it needs to search the casadi directory to find the lib, you could either do this by setting LD_LIBRARY_PATHS appropriately, or you could set the RPath for the `idaklu.***.so` file within the CMakeLists.txt file. I would try the latter approach first since you already have the correct CASADI_DIR (since cmake was able to link against the lib in the first place), you need to do something like:

```cmake
target_link_libraries(idaklu PRIVATE ${SUNDIALS_LIBRARIES} casadi) #after this line
set_property(TARGET idaklu APPEND PROPERTY INSTALL_RPATH "${CASADI_DIR}/lib")
```

note, not sure if "${CASADI_DIR}/lib" is even the right subdirectory!

note that I'm not terrible familiar with how to set rpath appropriately, or what cmake does by default, but this is my best guess. If this doesn't work perhaps try setting LD_LIBRARY_DIR appropriately during the fixing wheel step

Reply:
I fixed it on macOS. I could not get it to work within CMakeLists.txt like you suggested but it turns out that we needed to remove the rpath for `libc++.1.0.dylib` in `libcasadi.3.7.dylib` as well and not only `libcasadi.dylib` (confirmed it with `otool`, both of them had a shared dependency on it). I wonder if we can do the same fixes for GNU/Linux as well because it technically is the same error with `auditwheel` there...

Another solution I thought about (but haven't tested yet, will need to do so in CI) is to override the self-referenced `@rpath` for `libcasadi.3.7.dylib` itself and make it point towards itself with a fixed path, because the rpath will be fixed and installed into `pybamm/.dylibs/` anyway when `cibuildwheel` at the time of repair

Reply:
Resolving this as I manually appended casadi's installation path to LD_LIBRARY_PATH (I could not fix it through CMake no matter how I tried experimenting with CMAKE_INSTALL_RPATH and BUILD_RPATH) which fixed the issue and pointed `auditwheel` to `libcasadi.so.3.7`